### PR TITLE
Http branch

### PR DESCRIPTION
**Pull request by MohamedShatry**
Contains a lot of changes to handle nasty async stuff

### CODE REVIEW

File: src/main/ngapp/src/app/app-routing.module.ts [hunk 1]
```diff
@@ -11,15 +11,17 @@ const routes: Routes = [
   {
     path: 'login',
     component: LoginComponent,
-    canActivate: [AuthGuard],
+    canActivate: [AuthGuard]
   },
   {
     path: 'register',
     component: RegisterComponent,
+    pathMatch: 'full'
```
Comment (kiprot) [inline src/main/ngapp/src/app/app-routing.module.ts:19]: why do we have ```pathmatch: full``` here? Should it not just be on the chat path since it is the default landing page if a user is logged in?

File: src/main/ngapp/src/app/auth/auth.service.ts [hunk 2]
```diff
@@ -17,25 +19,26 @@ export class AuthService {
...
-    });
+  constructor(public afAuth: AngularFireAuth, private router: Router, private http: HttpClient) {
+    // this.afAuth.authState.subscribe((user) => {
+    //   if (user) {
+    //     this.user = user;
+    //     localStorage.setItem('user', JSON.stringify(this.user));
+    //   } else {
+    //     localStorage.setItem('user', null);
+    //   }
+    // });
```
Comment (kiprot) [inline src/main/ngapp/src/app/auth/auth.service.ts:32]: why was this deleted? shouldnt we be checking the authentification state?

File: src/main/ngapp/src/app/chat-data.service.ts [hunk 3]
```diff
@@ -1,44 +1,50 @@
...
+  // Create a new user object for the user.
+  addUser(user: User): void{
+      this.http.post("/user", user);
   }
 
-  addMessage(input: Message): void {
-    MESSAGES.push(input);
+  getUsers(): Observable<User[]> {
+    var localUser = JSON.parse(localStorage.getItem("user"));
+    return this.http.get<User[]>("/users?userId="+localUser.uid);
```
Comment (kiprot) [inline src/main/ngapp/src/app/chat-data.service.ts:34]: will this always be successful? i thought it is an asynchronous call?

File: src/main/ngapp/angular.json [hunk 4]
```diff
@@ -13,7 +13,8 @@
         "build": {
           "builder": "@angular-devkit/build-angular:browser",
           "options": {
-            "outputPath": "../webapp/ng",
+            "outputPath": "../webapp/",
+            "deleteOutputPath": false,
```
Comment (kiprot) [inline src/main/ngapp/angular.json:17]: what does deleting output path do?

File: src/main/ngapp/src/app/app-routing.module.ts (hunk 1 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/app-routing.module.ts:19]: It basically just ensures that we get that whole url. It doesn't actually affect much

File: src/main/ngapp/src/app/auth/auth.service.ts (hunk 2 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/auth/auth.service.ts:32]: I decided to manually add the auth state to localStorage whenever a user logsin/signs up/ registers

File: src/main/ngapp/src/app/chat-data.service.ts (hunk 3 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/chat-data.service.ts:34]: The parsing isn't asynchronous. Adding to local storage is what is asynchronous. Subscribing to the authentication state is what it asynchronous. That's why I changed it above

File: src/main/ngapp/src/app/auth/auth.service.ts (hunk 2 above)
Comment (kiprot) [reply src/main/ngapp/src/app/auth/auth.service.ts:32]: so the authState just lasts for the duration that the application tab/window is open and disappears when it is closed?

File: src/main/ngapp/angular.json (hunk 4 above)
Comment (MohamedShatry) [reply src/main/ngapp/angular.json:17]: It just ensures we don't delete the contents of the folder the app is built in. This way, we don't delete the xml file in the webapp directory

File: src/main/ngapp/src/app/auth/auth.service.ts (hunk 2 above)
Comment (kiprot) [reply src/main/ngapp/src/app/auth/auth.service.ts:32]: never mind. i think i got it

**Comment by kiprot** [comment]
Looks good to me
